packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts

```diff
@@ -400,40 +396,39 @@ export class GutenbergEditorPage {
 	 * @param {boolean} visit Whether to then visit the page.
 	 * @returns {Promise<string>} The published URL.
 	 */
-	async publish( { visit = false }: { visit?: boolean } = {} ): Promise< string > {
+	async publish( {
+		update = false,
+		visit = false,
+	}: { update?: boolean; visit?: boolean } = {} ): Promise< URL > {
 		const frame = await this.getEditorFrame();
 
-		const initialPublishButton = `${ selectors.publishButton(
-			selectors.postToolbar
-		) }, ${ selectors.scheduleButton( selectors.postToolbar ) }`;
-		const secondaryPublishButton = `${ selectors.publishButton(
-			selectors.publishPanel
-		) }, ${ selectors.scheduleButton( selectors.publishPanel ) }`;
-
-		await frame.click( initialPublishButton );
-		await frame.click( secondaryPublishButton );
-		const publishedURL = await this.getPublishedURL();
-
-		if ( visit ) {
-			await this.visitPublishedPost( publishedURL );
+		if ( update ) {
+			await frame.click( selectors.updateButton );
+		} else {
+			// The target button can have either one of the two strings:
+			// 	- Schedule
+			// 	- Publish
+			const initialPublishButton = `${ selectors.publishButton(
+				selectors.postToolbar
+			) }, ${ selectors.scheduleButton( selectors.postToolbar ) }`;
+
+			// Publishing a new article is a two-step process.
+			await frame.click( initialPublishButton );
+			await this.editorPublishPanelComponent.publish();
 		}
-		return publishedURL;
-	}
 
-	/**
-	 * Updates the post or page.
-	 *
-	 * @param {boolean} visit Whether to then visit the page.
-	 * @returns {Promise<string>} URL of the update post or page.
-	 */
-	async update( { visit = false }: { visit?: boolean } = {} ): Promise< string > {
-		const frame = await this.getEditorFrame();
-
-		await frame.click( selectors.updateButton );
-		const publishedURL = await this.getPublishedURL();
+		// In some cases the post may be published but the EditorPublishPanelComponent
+		// is either not present or forcibly dismissed due to a bug.
+		// eg. publishing a post with some of the Jetpack Earn blocks.
+		// By racing the two methods of obtaining the published article's URL, we can
+		// guarantee that one or the other works.
+		const publishedURL: URL = await Promise.race( [
+			this.editorPublishPanelComponent.getPublishedURL(),
+			this.getPublishedURLFromToast(),
+		] );
```

Comment:
As noted in the comment - our editor sometimes has a peculiar issue such as https://github.com/Automattic/wp-calypso/issues/50302 where the EditorPublishPanelComponent is either not present, or forcibly dismissed when publishing/updating/scheduling a post. By racing the promises we can obtain the published article URL even in that case.